File: ShellBoost.Samples/ShellBoost.Samples.Overview/SimpleFolder.cs
using System.Collections.Generic;
using System.Linq;
using ShellBoost.Core;
using ShellBoost.Core.WindowsShell;

namespace ShellBoost.Samples.Overview
{
    public class SimpleFolder : ShellFolder
    {
        public SimpleFolder(SimpleFolder parent, string name)
            : base(parent, new StringKeyShellItemId(name))
        {
            Level = parent.Level + 1;
        }

        // constructor reserved for root
        public SimpleFolder(ShellItemIdList idList)
            : base(idList)
        {
            // Level = 0
        }

        public new OverviewShellFolderServer FolderServer => (OverviewShellFolderServer)base.FolderServer;

        public int Level { get; }

        // set this to true to see images (of cute cats)
        public bool ShowImages => true;

        public override IEnumerable<ShellItem> EnumItems(SHCONTF options)
        {
            // add folders
            // note in this sample we only add folders up to two levels
            var maxLevels = 2;
            if (options.HasFlag(SHCONTF.SHCONTF_FOLDERS) && Level <= maxLevels)
            {
                var maxFolders = 2;
                for (var i = 0; i < maxFolders; i++)
                {
                    yield return new SimpleFolder(this, "Virtual Folder " + Level + "." + i);
                }
            }

            // add items
            if (options.HasFlag(SHCONTF.SHCONTF_NONFOLDERS))
            {
                var maxItems = 2;
                if (ShowImages)
                {
                    maxItems *= 2;
                    var i = 0;
                    for (; i < maxItems / 2; i++)
                    {
                        yield return new SimpleItem(this, "Virtual Item #" + i + ".txt");
                    }

                    for (; i < maxItems; i++)
                    {
                        var imgKey = (ulong)(Level * 10 + i);
                        yield return new SimplePngItem(this, "Virtual Image Key#" + imgKey + ".png", imgKey);
                    }
                }
                else
                {
                    for (var i = 0; i < maxItems; i++)
                    {
                        yield return new SimpleItem(this, "Virtual Item #" + i + ".txt");
                    }
                }
            }
        }

        protected override void MergeContextMenu(ShellFolder folder, IReadOnlyList<ShellItem> items, ShellMenu existingMenu, ShellMenu appendMenu)
        {
            // only with selected items
            if (items.OfType<SimplePngItem>().Count() > 0)
            {
                var clearLocal = new ShellMenuItem(appendMenu, "Clear Cache");
                clearLocal.Invoke += (s, e) =>
                {
                    foreach (var item in e.Items.OfType<SimplePngItem>())
                    {
                        item.ClearCache();
                    }
                };

                appendMenu.Items.Add(clearLocal);
            }
        }
    }
}

File: ShellBoost.Samples/ShellBoost.Samples.CppOverview/SimpleFolder.cpp
#include "pch.h"
#include "SimpleFolder.h"
#include "SimpleItem.h"
#include "SimplePngItem.h"

using namespace System::Linq;

SimpleFolder::SimpleFolder(SimpleFolder^ parent, String^ text)
	:ShellFolder(parent, gcnew StringKeyShellItemId(text))
{
	Level = parent->Level + 1;
	ShowImages = true;
}

SimpleFolder::SimpleFolder(ShellItemIdList^ idList)
	:ShellFolder(idList)
{
	// Level = 0
	ShowImages = true;
}

IEnumerable<ShellItem^>^ SimpleFolder::EnumItems(SHCONTF options)
{
	// real IEnumerable/yield is not a free lunch on C++/CLI, so we use a list
	auto list = gcnew List<ShellItem^>();

	// add folders
	// note in this sample we only add folders up to two levels
	auto maxLevels = 2;
	if (options.HasFlag(SHCONTF::SHCONTF_FOLDERS) && Level <= maxLevels)
	{
		auto maxFolders = 2;
		for (auto i = 0; i < maxFolders; i++)
		{
			list->Add(gcnew SimpleFolder(this, "Virtual Folder " + Level + "." + i));
		}
	}

	// add items
	if (options.HasFlag(SHCONTF::SHCONTF_NONFOLDERS))
	{
		auto maxItems = 2;
		if (ShowImages)
		{
			maxItems *= 2;
			auto  i = 0;
			for (; i < maxItems / 2; i++)
			{
				list->Add(gcnew SimpleItem(this, "Virtual Item #" + i + ".txt"));
			}

			for (; i < maxItems; i++)
			{
				auto  imgKey = (UInt64)(Level * 10 + i);
				list->Add(gcnew SimplePngItem(this, "Virtual Image Key#" + imgKey + ".png", imgKey));
			}
		}
		else
		{
			for (auto i = 0; i < maxItems; i++)
			{
				list->Add(gcnew SimpleItem(this, "Virtual Item #" + i + ".txt"));
			}
		}
	}

	return list;
}

ref class InvokeEventHandler
{
public:
	void Handler(Object^ sender, ShellMenuInvokeEventArgs^ e)
	{
		auto enumerator = Enumerable::OfType<SimplePngItem^>(e->Items)->GetEnumerator();
		while (enumerator->MoveNext())
		{
			enumerator->Current->ClearCache();
		}
	}
};

void SimpleFolder::MergeContextMenu(ShellFolder^ folder, IReadOnlyList<ShellItem^>^ items, ShellMenu^ existingMenu, ShellMenu^ appendMenu)
{
	if (Enumerable::Count(Enumerable::OfType<ShellItem^>(items)) > 0)
	{
		auto clearLocal = gcnew ShellMenuItem(appendMenu, "Clear Cache");
		auto handler = gcnew InvokeEventHandler();
		clearLocal->Invoke += gcnew EventHandler<ShellMenuInvokeEventArgs^>(handler, &InvokeEventHandler::Handler);

		appendMenu->Items->Add(clearLocal);
	}
}
